Generic BuildQueue

tgraf opened the pull request:
This introduces a basic build queue for objects identified by UUID. An object
is buildabe if it implements GetUUID(). An object can build itself if it
implements a Build() method.

Objects with identical UUID will never be built in parallel. The build queue
will maintain a set of worker threads which will build objects in parallel.

When an object is already queued to be built not being being built yet, any
further build request will be ignored.

When an object is being actively built and a new build request is enqueued, the
build request will deferred. It wil be scheduled into the build queue when the
currently ongoing build of that same object has completed.

<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cilium/cilium/5446)
<!-- Reviewable:end -->

eloycoto commented on pkg/endpoint/builder.go:
```diff
@@ -0,0 +1,73 @@
+func (b *endpointBuild) Build() error {
+	e := b.endpoint
+
+	scopedLog := e.getLogger()
```
I need to say that if you add in the `e.GetLogger` the UUID for the build you'll make me super happy! So, ops people can grep for the given UUID and will get all the information. 

Maybe add to `b.Context` a logrus.Entry with the uuid, and e.regenerate feed the logs from there. it will be great!

tgraf replied:
We might as well just make UUID part of the endpoint logger to have unique correlation

aanm commented on pkg/buildqueue/buildqueue.go:
```diff
@@ -0,0 +1,360 @@
+		// Wait for all regular builds to complete
+		q.ongoingRegularBuilds.Wait()
+
+		err := b.Build()
```
I believe with the current code, this build is not guarantee to be exclusive. Having a sleep in [0] will trigger what I'm saying.

Why don't you replace both `WaitGroup` with a `RWMutex` and use the `Lock` for the `ongoingExclusiveBuilds` and the `RLock` for the `ongoingRegularBuilds`? Oh because you want the others to finish before triggering this one... Does that really matter?

tgraf replied:
The whole point is to avoid holding mutexes for such a long time. This is still marked as wip. I'm not sure why everybody is reviewing this yet ;-) There are several other issues outstanding as well.

aanm replied:
>I'm not sure why everybody is reviewing this yet ;-) 

It's an exciting new world!

tgraf replied:
OK, OK. Just hold your horses for a little bit longer ;-)

tgraf replied:
I've added more extensive testing with random build durations plus constant checking of build guarantees to provide adequate test coverage.

aanm commented on pkg/buildqueue/buildqueue.go:
```diff
@@ -0,0 +1,360 @@
+
+		// Cannot run regular builds while exclusive builds are ongoing
+		q.ongoingExclusiveBuilds.Wait()
+
```
[0] `time.Sleep(5*time.Second)`

aanm commented on pkg/endpoint/endpoint.go:
```diff
@@ -461,6 +462,9 @@ type Endpoint struct {
 	egressPolicyEnabled bool
 
+	// uuid is the UUID of the endpoint. It is assigned via GetUUID()
+	uuid string
```
Why `uuid` for the endpoint? We already have the `id`. More concerns left in the `GetUUIDLocked()`

aanm commented on pkg/endpoint/endpoint.go:
```diff
@@ -2742,3 +2748,17 @@ func (e *Endpoint) scrubIPsInConntrackTable() {
+
+func (e *Endpoint) GetUUIDLocked() string {
+	if e.uuid == "" {
+		e.uuid = uuid.NewUUID().String()
```
Because you are generating the NewUUID based on something random. I would prefer the "real" uuid to be something like `fmt.Sprintf("%s-%s-%s", ClusterName, NodeName, EndpointID)`

tgraf replied:
endpoint ID is not unique over time. The same endpoint ID can be reused. What's wrong with using a guaranteed unique UUID for these purposes?

aanm replied:
>endpoint ID is not unique over time. 

that's true. I forgot about this

>What's wrong with using a guaranteed unique UUID for these purposes?
It's just that we don't have any correlation between the UUID and where the endpoint is which thinking more about it, it doesn't really matter.

eloycoto commented on pkg/metrics/metrics.go:
```diff
@@ -308,6 +326,14 @@ var (
+
+	// BuildQueueEntries is the number of queued, waiting and running
+	// builds in the build queue
+	BuildQueueEntries = prometheus.NewGaugeVec(prometheus.GaugeOpts{
```
Hey, you didn't add this one into `Documentation/config/metrics.rst`

eloycoto commented on pkg/endpoint/builder.go:
```diff
@@ -0,0 +1,100 @@
+}
+
+func (b *endpointBuild) GetUUID() string {
+	return b.endpoint.GetUUID()
```
Umm, I'm not agree in this one, each build should have an specific build ID no?  What happens if one endpoint have two builds waiting? Both will be with the same UUID and there is no way to know the status, and on delete from queue maybe are some conflicts. 

I would not use the UUID in the endpoint and I'll use a new one per each build, as do other Queue jobs managers like Celery, Airflow or Machinery

tgraf replied:
The UUID does not describe the build. It describes the buildable. If the same buildable is being built multiple times, all of these builds will have the same UUID and are correlated as such. We previously used the endpoint.ID for this purpose but the endpoint.ID is fragile as it can be reused again. Introducing a UUID allows grepping for the UUID in the logs.

eloycoto commented on pkg/endpoint/builder.go:
```diff
@@ -0,0 +1,100 @@
+	e := b.endpoint
+
+	scopedLog := e.Logger()
+	scopedLog.Debug("Dequeued endpoint from build queue")
```
I would change this one to `starting building endpoint` make more sense for users.

tgraf replied:
This is an existing log message. I had just moved it. I've just removed it now as `regenerate()` will print a verbose message as well and it's really a weak duplicate.

eloycoto commented on pkg/endpoint/builder.go:
```diff
@@ -0,0 +1,100 @@
+	scopedLog := e.Logger()
+	scopedLog.Debug("Dequeued endpoint from build queue")
+
+	err := e.regenerate(b.owner, b.context)
```
Quick one, if the endpoint takes a few minutes in the queue, we'll lost the total time duration because queue time no? I think that metrics should change to something like this: 
 
```
diff --git a/pkg/endpoint/builder.go b/pkg/endpoint/builder.go
index 94ee34e31..1138978fe 100644
--- a/pkg/endpoint/builder.go
+++ b/pkg/endpoint/builder.go
@@ -33,6 +33,7 @@ type endpointBuild struct {
 }
 
 func (e *Endpoint) newEndpointBuild(owner Owner, context *RegenerationContext) *endpointBuild {
+	context.Stats.queueTime.Start()
 	return &endpointBuild{
 		endpoint: e,
 		owner:    owner,
@@ -62,7 +63,7 @@ func (b *endpointBuild) BuildsDequeued(nbuilds int, cancelled bool) {
 
 func (b *endpointBuild) Build() error {
 	e := b.endpoint
-
+	b.context.Stats.queueTime.End()
 	scopedLog := e.Logger()
 	scopedLog.Debug("Dequeued endpoint from build queue")
 
diff --git a/pkg/endpoint/metrics.go b/pkg/endpoint/metrics.go
index deaca8849..87b678283 100644
--- a/pkg/endpoint/metrics.go
+++ b/pkg/endpoint/metrics.go
@@ -25,6 +25,7 @@ import (
 
 type regenerationStatistics struct {
 	success                bool
+	queueTime              spanstat.SpanStat
 	totalTime              spanstat.SpanStat
 	waitingForLock         spanstat.SpanStat
 	waitingForCTClean      spanstat.SpanStat
diff --git a/pkg/endpoint/policy.go b/pkg/endpoint/policy.go
index a6fbf4613..50f159ecf 100644
--- a/pkg/endpoint/policy.go
+++ b/pkg/endpoint/policy.go
@@ -680,11 +680,9 @@ func (e *Endpoint) regenerate(owner Owner, context *RegenerationContext) (retErr
 	var compilationExecuted bool
 	var err error
 
-	context.Stats = regenerationStatistics{}
 	stats := &context.Stats
 
 	metrics.EndpointCountRegenerating.Inc()
-	stats.totalTime.Start()
 	e.Logger().WithFields(logrus.Fields{
 		logfields.StartTime: time.Now(),
 		logfields.Reason:    context.Reason,
@@ -797,6 +795,13 @@ func (e *Endpoint) regenerate(owner Owner, context *RegenerationContext) (retErr
 // either StateWaitingForIdentity if the security identity is still unknown or
 // to StateReady.
 func (e *Endpoint) Regenerate(owner Owner, context *RegenerationContext) <-chan bool {
+
+	context.Stats = regenerationStatistics{}
+	stats := &context.Stats
+
+	metrics.EndpointCountRegenerating.Inc()
+	stats.totalTime.Start()
+
 	return BuildQueue.Enqueue(e.newEndpointBuild(owner, context))
 }

```

So, in this way we can know how long was waiting in the queue and the total endpoint time is correct.

tgraf replied:
Very good point. I've added this. I didn't change the meaning of the regenerating metric though as the buildqueue has its own metric which is more fine grained.

ianvernon commented on pkg/endpoint/endpoint.go:
```diff
@@ -461,6 +481,27 @@ type Endpoint struct {
+	// UUID is the UUID of the endpoint.
+	UUID string
+
+	// initialBuildSuccessful is true after the first build has been
```
I'd prefer we encode data about the 'build' of the endpoint in its own structure vs. having separate fields directly within endpoint. This could include the state as well.

tgraf replied:
I've moved it into a struct `buildStatus`. It does not change that the endpoint is currently implementing the `Buildable`. Do you have a higher level concern that we store state in the endpoint that is relative to whether the endpoint has been built?

I've separated it like this:
- Information and state regarding the currently ongoing build is in `endpointBuilder` and in the `pkg/buildqueue` package
- Information regarding the state of the endpoint continues to be in the endpoint. This includes whether the endpoint has been successfully built once, whether the last build failed, whether the endpoint is currently being built, etc. This *does not* relate to a particular build. It is the overall state considering all ongoing builds for that endpoint.

Let me know if that makes sense to you.

ianvernon commented on pkg/endpoint/endpoint.go:
```diff
@@ -2739,3 +2630,66 @@ func (e *Endpoint) scrubIPsInConntrackTable() {
+
+// GetUUID returns the UUID of the endpoint
+func (e *Endpoint) GetUUID() string {
+	return e.UUID
```
Why have a getter for UUID if the field is public?

tgraf replied:
Removed

tgraf replied:
EDIT: I forgot that this was added to implement the `Buildable` interface. I've added it back and added a comment to mention the interface use.

ianvernon commented on daemon/endpoint.go:
```diff
@@ -425,9 +425,6 @@ func (d *Daemon) deleteEndpointQuiet(ep *endpoint.Endpoint, releaseIP bool) []er
 
```
Typo in commit message: remoev --> remove

ianvernon commented on pkg/endpoint/builder.go:
```diff
@@ -33,6 +33,7 @@ type endpointBuild struct {
 }
 
 func (e *Endpoint) newEndpointBuild(owner Owner, context *RegenerationContext) *endpointBuild {
+	context.Stats.queueWait.Start()
```
If anyone ever passes a nil `RegenerationContext`, this may panic.

tgraf replied:
We already rely on context not being nil across the entire regeneration code. What do you suggest to do if it is nil? Do you want to add it to all functions?

ianvernon replied:
Given that we control the value of `context`, this shouldn't be an issue. I was speaking from prior experience with dereferencing of pointers directly, d2c0887afc8843e9c38df1da24ef53eca67dbd53 .

ianvernon reviewed (commented):
Took a first pass at the new changes, I will look over again because this is a huge change and thus need to "digest" a bit before looking at it again.

aanm commented on Documentation/concepts.rst:
```diff
@@ -324,6 +324,47 @@ example is a Kubernetes cluster which uses containerd as the container runtime.
+Endpoint States
+---------------
+
+An endpoint can be in the following state to describe its current status:
```
Where's the `not-ready`?

tgraf replied:
Interestingly it's a state unknown to the Cilium internals. We currently label endpoints as not-ready even if a previous build succeeded and the endpoint is connected. I'm planning to remove it but I didn't want to overload this PR any further. What is now `not-ready` is completely covered by the combination of `creating`, `restoring`, `init`,  and `outdated`

jrajahalme replied:
For as long as I can remember, 'not-ready' has been a state visible via the API, but it was never defined as one of the endpoint states below the API. I have no idea if there is a good reason for this, other than maybe a desire to simplify the state visible via the API without complicating the endpoint's internal state with it?

aanm commented on pkg/endpoint/builder.go:
```diff
@@ -0,0 +1,115 @@
+
+	// numWaiting is the number of builds of this endpoint waiting to be
+	// built
+	numWaiting int
```
I don't understand the description of this variable. Why would an endpoint have more than 1 build waiting to be built? If the endpoint is queued multiple times to be build, won't all of the queue requests be dropped as the endpoint is already in place to be build?

tgraf replied:
They won't be dropped, they will be folded into the already queued build and all builds will be notified for completion afterward. Extended comment to make this clearer.

tgraf replied:
I've extended the comment to make it crystal clear.

aanm commented on pkg/endpoint/builder.go:
```diff
@@ -0,0 +1,115 @@
+}
+
+func (b *endpointBuild) BuildsDequeued(nbuilds int, cancelled bool) {
+	if err := b.endpoint.LockAlive(); err == nil {
```
If an endpoint is in disconnecting state we won't dequeue it?

tgraf replied:
The build will be dequeued but `regenerate()` will fail as the endpoint is no longer alive. You raise a good point here though that it still makes sense to maintain `numWaiting`. It doesn't affect the endpoint state anymore as we can't transition out of disconnecting state other than to become disconnected but it is more correct to make this an unconditional lock. Fixed.

aanm commented on pkg/endpoint/policy.go:
```diff
@@ -667,6 +667,13 @@ func (e *Endpoint) ComputePolicyEnforcement(repo *policy.Repository) (ingress bo
+// WaitForRegenerationsToComplete waits until any ongoing current build has
+// completed and removes all scheduled builds from the build queue.  The
+// endpoint must NOT be locked.
+func (e *Endpoint) WaitForRegenerationsToComplete() {
```
I would rename this function. I honestly thought this was waiting for the endpoint regeneration to be completed (non-destructive) but it's actually draining (destructive operation) the endpoint from the queue.

tgraf replied:
Good point, renamed.

aanm commented on pkg/buildqueue/buildqueue.go:
```diff
@@ -0,0 +1,542 @@
+	close(b.internalNotification)
+
+	for _, notify := range b.notificationChannels {
+		notify <- success
```
if the reader never reads the channel this will block forever

tgraf replied:
It should not, the channel is buffered with a length of 1 and is guaranteed to only be written to once so the answer is waiting in the channel to be read and it is up to the caller to read the channel or ignore it. Or am I missing something @aanm?

aanm commented on pkg/buildqueue/buildqueue.go:
```diff
@@ -0,0 +1,542 @@
+// addNotificationChannel adds channels to be notified. This function may only
+// be called while the build is not yet in progress.
+func (b *build) addNotificationChannel(n BuildNotification) {
+	b.notificationChannels = append(b.notificationChannels, n)
```
consider having a function that returns a notification channel. `getNotificationChannel`. By creating the channel returned in this function you know that you can write and close the channel without any worries on line 229.

tgraf replied:
👍 The function already existed in the form of `newBuildNotification()`. I've consolidated it into a single function.

aanm commented on pkg/buildqueue/buildqueue.go:
```diff
@@ -0,0 +1,542 @@
+	name string
+
+	// mutex protects the queued map
+	mutex lock.Mutex
```
Can you put the mutex and the `buildStatusMap` at the end of the structure? At first I thought `mutex` was protecting all attributes of this `struct` except name.

tgraf replied:
The mutex is also used for the build condition. I've added documentation.

aanm commented on pkg/buildqueue/buildqueue.go:
```diff
@@ -0,0 +1,542 @@
+
+	defer buildStatus.builder.BuildsDequeued(num, false)
+
+	q.mutex.Unlock()
```
Why isn't this on line 447?

tgraf replied:
Good point, fixed

aanm reviewed (commented):
I'll just leave my 2 cents, but I can't explicitly approve it or request changes.

Unfortunately I can't understand the logic behind the code as there are terms that seem logically similar to me, e.g. "queue" / "waitForSlot" but they have different behaviors, e.g. doing a `for b := range q.workerBuildQueue` is, at least for me, a "dequeue" of `b` from `q.workerBuildQueue` but a couple lines bellow the current line is executed `currentBuild := q.dequeueBuild(uuid)` which implies `b` is going to be dequeued again.

I'm all in favor for the changes but I have to be honest that I personally won't be able to maintain this code in the future.

rlenglet commented on pkg/metrics/metrics.go:
```diff
@@ -33,6 +33,18 @@ import (
 )
 
+const (
+	// BuildQueueWaiting is the number of entries waiting in the build queue
```
Clarify that those constants are used as values in vector metrics:
```
... is the state of a build that is waiting in a build queue.
```

rlenglet replied:
Rename those constants. Those are not about _queues_, they are about _build_ states:
`BuildQueueWaiting` -> `BuildStateWaiting`
etc.

rlenglet commented on pkg/metrics/metrics.go:
```diff
@@ -80,6 +92,12 @@ var (
 	// (scope=slow_path)
 	LabelScope = "scope"
 
+	// LabelBuildQueueState is the state a build queue entry is in
```
The description is correct, but the name implies that this is the state of a _queue_ instead of a _build_. Rename accordingly:
`LabelBuildQueueState` -> `LabelBuildState`

rlenglet commented on pkg/metrics/metrics.go:
```diff
@@ -359,6 +385,8 @@ func init() {
 
 	MustRegister(ErrorsWarnings)
+
+	MustRegister(BuildQueueEntries)
```
Nitpick: `Register.MustRegister` takes a slice of collectors. I think this would be a bit more readable, and just as maintainable:
```
MustRegister(
    prometheus.NewProcessCollector(os.Getpid(), Namespace)),
    EndpointCountRegenerating,
    ...
    BuildQueueEntries,
)
```

tgraf replied:
I agree, but I did not want to refactor existing code in this PR outside of introducing the build queue. Happy if somebody does this afterward. It's unrelated to my change. I'm keeping the existing style.

rlenglet commented on pkg/buildqueue/buildqueue.go:
```diff
@@ -0,0 +1,568 @@
+	// waitingBuilders sync.Condition in waitForBuildCondition()
+	mutex lock.Mutex
+
+	// buildStatus contains the status of all builders separated by UUID
```
`separated` -> `indexed`

rlenglet commented on pkg/buildqueue/buildqueue.go:
```diff
@@ -0,0 +1,568 @@
+	// builds
+	workerBuildQueue chan Builder
+
+	// stopWorker is used to stop worker threads when the build queue is
```
`worker threads` -> `workers and exclusive builders`

rlenglet commented on pkg/buildqueue/buildqueue.go:
```diff
@@ -0,0 +1,568 @@
+	stopWorker chan struct{}
+
+	// regularBuilds accounts for running regular builds
+	regularBuilds *buildGroup
```
This doesn't need to be a pointer. Same for `exclusiveBuilds` below.

rlenglet commented on pkg/buildqueue/buildqueue.go:
```diff
@@ -0,0 +1,568 @@
+	Buildable
+
+	// Build must build object
+	Build() error
```
Nitpick: reorder the methods to match the order in which they are called:
- `BuildQueued`
- `BuildsDequeued`
- `Build`

I would also document that order in the type's doc above.
Also document that both `BuildQueued` and `BuildsDequeued` are always called once a builder is enqueued. But `Build` may not be called, e.g. if it's removed or drained.

rlenglet commented on pkg/buildqueue/buildqueue.go:
```diff
@@ -0,0 +1,568 @@
+	// about the completion of the build
+	notificationChannels []BuildNotification
+
+	// internalNotification is used to notify
```
to notify what?

rlenglet commented on pkg/buildqueue/buildqueue.go:
```diff
@@ -0,0 +1,568 @@
+		builder:              b,
+		uuid:                 uuid,
+		notificationChannels: []BuildNotification{},
+		internalNotification: make(BuildNotification, 1),
```
For consistency:
```
internalNotification: newBuildNotification(),
```

rlenglet commented on pkg/buildqueue/buildqueue.go:
```diff
@@ -0,0 +1,568 @@
+		metrics.LabelBuildQueueState: metrics.BuildQueueWaiting,
+	}).Inc()
+
+	b.BuildQueued()
```
Since the `BuildsDequeued()` and `Build()` callbacks are always made on the first queued `Builder` instance, for consistency I would also make the `BuildQueued()` callbacks on the same instance:
```
build.BuildQueued()
```
This would simplify managing the state of a set of builders for users in case they queue multiple instances of `Builder` with the same UUID.

rlenglet commented on pkg/buildqueue/buildqueue.go:
```diff
@@ -0,0 +1,568 @@
+// * In the presence of preemptive, exclusive builds, regular builds get a fair
+//   chance to run but preemptive builds always preempt any queued, not yet
+//   running regular build.
+func (q *BuildQueue) Enqueue(b Builder) BuildNotification {
```
To provide more safety, I would return a read-only channel here, instead of the read-write `BuildNotification`.

rlenglet commented on pkg/buildqueue/buildqueue.go:
```diff
@@ -0,0 +1,568 @@
+	uuid := b.GetUUID()
+	q.mutex.Lock()
+	if status, ok := q.buildStatus[uuid]; ok {
+		delete(q.buildStatus, uuid)
```
The build might be in progress. Shouldn't we keep the builder in the map in that case, until the build completes? Or maybe it would even be simpler to keep the builder in the map if `currentBuild` is not `nil`, since it would be deleted anyway after the build completes?

I think it works now, since `needToBuildAgain()` handles the deletion from the map well, but it might be surprising. In the normal case, a build may be deleted from the map only after a build has completed and there is no need to rebuild.

tgraf replied:
Yes, I agree that is more correct. The build status is guaranteed to be cleared up after the build completes.

tgraf replied:
I have changed the code to keep the buildStatus if a build is ongoing and defer the deletion to `needToBuildAgain()`.

rlenglet commented on pkg/buildqueue/buildqueue.go:
```diff
@@ -0,0 +1,568 @@
+// running, the build will be folded into the already scheduled build but both
+// notification channels will be notified.
+func (q *BuildQueue) PreemptExclusive(b Builder) BuildNotification {
+	notify, enqueue := q.enqueueBuild(b)
```
`enqueue` -> `enqueued`

rlenglet commented on pkg/buildqueue/buildqueue.go:
```diff
@@ -0,0 +1,568 @@
+// If an exclusive build for the same UUID is already enqueued but not yet
+// running, the build will be folded into the already scheduled build but both
+// notification channels will be notified.
+func (q *BuildQueue) PreemptExclusive(b Builder) BuildNotification {
```
Same as for normal builds, I would return a read-only channel instead of a read-write `BuildNotification`.

rlenglet commented on pkg/buildqueue/buildqueue.go:
```diff
@@ -0,0 +1,568 @@
+	}
+}
+
+// checkBuildGuarantees reports errors when build guarantees are currently not
```
guarantees -> invariants (?)

tgraf replied:
I've changed it to build condition to be consistent in naming.

rlenglet reviewed (dismissed):
Reviewed the `Generic BuildQueue...` commit. Only relatively minor issues. Will review the rest of the commits ASAP.

rlenglet commented on Documentation/concepts.rst:
```diff
@@ -324,6 +324,50 @@ example is a Kubernetes cluster which uses containerd as the container runtime.
+  that represents the connectivity state.
+
+* **Init:** The endpoint has been successfully regenerated once but has not been
+  able to resolve its identity yet. The endpoint is connected and is using
```
`to resolve its identity` -> `to retrieve its metadata`

rlenglet commented on Documentation/concepts.rst:
```diff
@@ -324,6 +324,50 @@ example is a Kubernetes cluster which uses containerd as the container runtime.
+
+* **Init:** The endpoint has been successfully regenerated once but has not been
+  able to resolve its identity yet. The endpoint is connected and is using
+  the init policy.
```
```
..., and uses the reserved:init label as its identity.
```

rlenglet commented on pkg/endpoint/builder.go:
```diff
@@ -0,0 +1,121 @@
+	buildQueue = buildqueue.NewBuildQueue("endpoint-builder")
+)
+
+type endpointBuild struct {
```
`endpointBuild` -> `endpointBuilder`

rlenglet commented on pkg/endpoint/builder.go:
```diff
@@ -0,0 +1,121 @@
+	numWaiting int
+}
+
+func (e *Endpoint) newEndpointBuild(owner Owner, context *RegenerationContext) *endpointBuild {
```
`newEndpointBuild` -> `newEndpointBuilder`

rlenglet commented on pkg/endpoint/builder.go:
```diff
@@ -0,0 +1,121 @@
+		}
+
+		e.buildStatus.building = false
+		e.buildStatus.lastFailed = err != nil
```
There might be a confusion, I would expect that `err` here is the error returned by `regenerate()`, but it's the error returned by `e.LockAlive()`. Is that intentional?

tgraf replied:
Good catch. This is a real bug. I've renamed all error variables to make this very clear.

rlenglet reviewed (commented):
The rest of the PR LGTM.

ianvernon commented on Documentation/configuration/metrics.rst:
```diff
@@ -39,6 +39,14 @@ Endpoint
 * ``endpoint_regeneration_time_stats_seconds``: Endpoint regeneration time stats labeled by scope.
```
Typo in commit - buildabe --> buildable

ianvernon commented on Documentation/configuration/metrics.rst:
```diff
@@ -39,6 +39,14 @@ Endpoint
+-----------
+
+* ``buildqueue_entries``: Number of queued, waiting, and running builds.
+  * ``state=running``: Number of builds currently being built
```
currently being built --> in progress

ianvernon commented on pkg/buildqueue/buildqueue.go:
```diff
@@ -0,0 +1,580 @@
+// Copyright 2016-2018 Authors of Cilium
```
I have no idea if this is correct or not (so feel free to disregard) - should this be 2018 only instead of 2016-2018?

tgraf replied:
I derived it from existing code so I preserved the original copyright header

ianvernon commented on pkg/buildqueue/buildqueue.go:
```diff
@@ -0,0 +1,580 @@
+type build struct {
+	builder Builder
+
+	// uuid is the UUID of the Buildable
```
Buildable --> Builder?

tgraf replied:
The UUID is derived from the Buildable interface

ianvernon commented on pkg/buildqueue/buildqueue.go:
```diff
@@ -0,0 +1,580 @@
+	return numBuilds
+}
+
+// numFolded returns the number of folded builds
```
This isn't blocking, but a lot of the function documentation in this PR is self-referential and thus doesn't provide much value. Given that this is not an exported function, it's not a big deal, though.

ianvernon commented on pkg/buildqueue/buildqueue.go:
```diff
@@ -0,0 +1,580 @@
+
+// numFolded returns the number of folded builds
+func (b *build) numFolded() int {
+	return len(b.notificationChannels)
```
Can you elaborate on how the number of folded builds relates to the number of notification channels ?

ianvernon commented on pkg/buildqueue/buildqueue.go:
```diff
@@ -0,0 +1,580 @@
+	return notify, enqueued
+}
+
+func (q *BuildQueue) serializeBuild(b Builder) (*build, bool) {
```
Document what the boolean represents here.

ianvernon commented on pkg/buildqueue/doc.go:
```diff
@@ -0,0 +1,28 @@
+// limitations under the License.
+
+// Package buildqueue provides a basic build queue for objects identified by
+// UUID. An object is buildabe if it implements GetUUID(). An object can build
```
buildable

ianvernon reviewed (commented):
Just minor documentation fixes, nothing blocking.

tgraf committed:
buildqueue: Introduce a generic object buildqueue

This introduces a basic build queue for objects identified by UUID. An object
is buildable if it implements GetUUID(). An object can build itself if it
implements a Build() method.

Objects with identical UUID will never be built in parallel. The build queue
will maintain a set of worker threads which will build objects in parallel.

When an object is already queued to be built not being being built yet, any
further build request will be ignored.

When an object is being actively built and a new build request is enqueued, the
build request will deferred. It wil be scheduled into the build queue when the
currently ongoing build of that same object has completed.

Signed-off-by: Thomas Graf <thomas@cilium.io>
